dkc/core/rest/authorized_upload.py

```diff
@@ -0,0 +1,106 @@
+
+from dkc.core.models import AuthorizedUpload, Folder
+from dkc.core.permissions import Permission
+
+
+class AuthorizedUploadSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = AuthorizedUpload
+        fields = ['id', 'created', 'creator', 'expires', 'folder', 'signature']
+
+
+class CreateAuthorizedUploadSerializer(serializers.Serializer):
+    folder = serializers.IntegerField()
+
+
+class CompleteAuthorizedUploadSerializer(serializers.Serializer):
+    authorization = serializers.CharField()
+
+
+class CanDeleteAuthorization(BasePermission):
+    message = 'Only the creator of an authorized upload may delete it.'
+
+    def has_object_permission(self, request: Request, view: View, obj: AuthorizedUpload) -> bool:
+        return view.action != 'destroy' or request.user.is_superuser or request.user == obj.creator
+
+
+class AuthorizedUploadViewSet(mixins.DestroyModelMixin, GenericViewSet):
+    queryset = AuthorizedUpload.objects.all()
+
+    permission_classes = [CanDeleteAuthorization]
+    serializer_class = AuthorizedUploadSerializer
+
+    @swagger_auto_schema(
+        responses={201: AuthorizedUploadSerializer},
+        request_body=CreateAuthorizedUploadSerializer,
+    )
+    def create(self, request, pk=None):
+        """Authorize an upload to a folder on your behalf."""
+        serializer = CreateAuthorizedUploadSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        folder = get_object_or_404(Folder, pk=serializer.validated_data['folder'])
+
+        if not folder.has_permission(request.user, Permission.write):
+            raise PermissionDenied('You are not allowed to create files in this folder.')
+
+        expires = timezone.now() + timedelta(days=settings.DKC_AUTHORIZED_UPLOAD_EXPIRATION_DAYS)
+        upload = AuthorizedUpload.objects.create(
+            folder=folder, creator=request.user, expires=expires
+        )
+        serializer = AuthorizedUploadSerializer(upload)
+        return Response(serializer.data, status=201)
+
+    @swagger_auto_schema(
+        responses={204: 'The authorized upload was finalized.'},
+        request_body=CompleteAuthorizedUploadSerializer,
+    )
+    @action(
+        detail=True, methods=['post'], queryset=AuthorizedUpload.objects.select_related('folder')
+    )
+    def completion(self, request, pk=None):
+        """Mark an authorized upload as complete and notify its originator."""
+        serializer = CompleteAuthorizedUploadSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        upload = self.get_object()
+
+        try:
+            upload.verify_signature(serializer.validated_data['authorization'])
+        except signing.BadSignature:
+            logger = logging.getLogger('django.security.SuspiciousOperation')
+            logger.warning('Authorized upload signature tampering detected.')
+            raise PermissionDenied('Invalid authorization signature.')
```

Comment:
A garbage signature is probably a 400 / `ValidationError`, rather than a 403 / `PermissionDenied`. If we put this in the serializer, this should get fixed naturally.